### PR TITLE
Initial project restructure

### PR DESCRIPTION
**Pull request by SlayterDev**
Took a stab at restructuring the generation scripts a bit. Now that we're getting more complex it will be helpful to modularize some areas.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -1,102 +1,53 @@
 const fs = require('fs')
-
-const OVERRIDE_DIR = 'overrides'
-const GENERATED_DIR = 'generated'
-const LISTS_DIR = 'features'
+const utils = require('./lib/utils')
```
Comment (jonathanKingston) [inline index.js:2]: This brings in a whole heap of require/export can we move to ESM like content-scope-scripts does? It should be just a case of adding _"type": "module"_ to the package.json as we're just a node script.

File: lib/constants.js [hunk 2]
```diff
@@ -0,0 +1,5 @@
+module.exports = {
+    OVERRIDE_DIR: 'overrides',
```
Comment (jonathanKingston) [inline lib/constants.js:2]: Moving to a constants file makes sense when they're used in multiple places, I think they're only in index.js right?

File: index.js [hunk 3]
```diff
@@ -1,102 +1,53 @@
...
 
 // Handle platform specific overrides and write configs to disk
 for (const platform of platforms) {
     const platformConfig = JSON.parse(JSON.stringify(defaultConfig))
-    const overridePath = `${OVERRIDE_DIR}/${platform}-override.json`
+    const overridePath = `${constants.OVERRIDE_DIR}/${platform}-override.json`
 
     if (!fs.existsSync(overridePath)) {
-        writeConfigToDisk(platform, platformConfig)
+        utils.writeConfigToDisk(platform, platformConfig)
```
Comment (jonathanKingston) [inline index.js:50]: This seems like a *existing* bug, we're not passing the second argument and we're creating the generated config files if the override doesn't exist?!
Probably can be deleted.

File: lib/utils.js [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+/**
+ * Write a config file to disk
+ *
+ * @param {string} platform - platform to write
+ * @param {object} config - the object to write
+ */
+function writeConfigToDisk (platform, config, v1Config) {
+    fs.writeFileSync(`${constants.GENERATED_DIR}/v2/${platform}-config.json`, JSON.stringify(config, null, 4))
+    fs.writeFileSync(`${constants.GENERATED_DIR}/v1/${platform}-config.json`, JSON.stringify(v1Config, null, 4))
```
Comment (jonathanKingston) [inline lib/utils.js:12]: This v1 saving would be better as part of the legacy generation, in hindsight the `v1Config` argument here seems like a mistake...

File: lib/legacy-gen/v1generator.js [hunk 5]
```diff
@@ -0,0 +1,19 @@
+function generateV1Config (platformConfig) {
```
Comment (jonathanKingston) [inline lib/legacy-gen/v1generator.js:1]: Both this and legacy could be implemented as transitions on the current format such that the code would just recursively call it and then there'd be less versioned code over time. v0 could special case the old files perhaps too.

```js
export const legacyFormats = [
   {
        configVersion: 'v1',
        transition(dataIn) {
            const dataOut = dataIn;
            ...
            return dataOut
        },
    }
    {
        configVersion: 'v0',
        ....
    }
]
```

```js
import { legacyFormats } from './legacyFormats.js'

function writeToDisk(currentFormat) {
    writeConfigToDisk(currentConfigVersion, platform, dataOut)
    for (let format of legacyFormats) {
        const dataOut = format.transition(currentFormat)
        if (format.configVersion === 'v0') {
            writeLegacyFilesToDisk(dataOut)
        } else {
            writeConfigToDisk(format.configVersion, platform, dataOut)
        }
        currentFormat = dataOut
    }
}
```

**Review by jonathanKingston — CHANGES_REQUESTED** [review]
I think there are bigger changes we can bring within this that are worth discussing... added my comments.

File: lib/constants.js (hunk 2 above)
Comment (SlayterDev) [reply lib/constants.js:2]: If we move to saving in the legacy generation they'll be used elsewhere. I think at least GENERATED is already used in a couple files.

File: lib/legacy-gen/v1generator.js (hunk 5 above)
Comment (SlayterDev) [reply lib/legacy-gen/v1generator.js]: I like the idea of the legacy transition structure but I think the v0 config files are just special enough, especially with the need for the unprotectedTemp object, that they should be kept as a separate case done at then end of generation. The others make sense to be done per platform file on write but the v0 files only need to be done once.

**Comment by jonathanKingston** [comment]
Let's add the genericising of the eol feature here too. Otherwise this looks good; I'll wait for the fledge change to drop before giving the final r+ here.

**Comment by SlayterDev** [comment]
Closing for now as this will need some refactoring after recent updates.

**Pull request closed without merging**